lang/src/org/partiql/lang/eval/ExprValueExtensions.kt

```diff
@@ -384,20 +395,19 @@ fun ExprValue.cast(
 
                             // Note that the [genericTimeRegex] has a group to extract the zone offset.
                             val zoneOffsetString = matcher.group(2)
-                            // Check if timezone is specified
-                            val zoneOffset = when(timezone) {
-                                0 -> null
-                                else -> zoneOffsetString?.let { ZoneOffset.of(it) } ?: LOCAL_TIMEZONE_OFFSET
-                            }
+                            val zoneOffset = zoneOffsetString?.let { ZoneOffset.of(it) } ?: LOCAL_TIMEZONE_OFFSET
 
-                            // Check if precision is specified or not. Value greater than MAX_PRECISION_FOR_TIME indicates that the precision was not specified
-                            // In this case infer the precision from the TIME string.
-                            val precision = when {
-                                precisionProvided > MAX_PRECISION_FOR_TIME -> getPrecisionFromTimeString(stringValue())
-                                else -> precisionProvided
-                            }
-                            return valueFactory.newTime(Time.of(localTime.hour, localTime.minute, localTime.second, localTime.nano, precision,
-                                zoneOffset?.totalSeconds?.div(SECONDS_PER_MINUTE)
+                            return valueFactory.newTime(
+                                Time.of(
+                                    localTime.hour,
+                                    localTime.minute,
+                                    localTime.second,
+                                    localTime.nano,
+                                    precision?: getPrecisionFromTimeString(stringValue()),
+                                    when (targetSqlDataType) {
+                                        SqlDataType.TIME_WITH_TIME_ZONE -> zoneOffset.totalSeconds / SECONDS_PER_MINUTE
+                                        else -> null
+                                    }
```

Comment:
I would extract this as a variable--it is confusing to nest it here.

Reply:
I see. Changed it.